### PR TITLE
Cold storage framework

### PR DESCRIPTION
**Pull request by nicolas2bert**
Not intended to be merged. For direction only.

### CODE REVIEW

File: extensions/lifecycle/LifecycleQueuePopulator.js [hunk 1]
```diff
@@ -117,6 +117,115 @@ class LifecycleQueuePopulator extends QueuePopulatorExtension {
...
+        const isObjectAlreadyRestored = value.archive
+        && value.archive.restoreCompletedAt
+        && new Date(value.archive.restoreWillExpireAt) >= new Date(Date.now());
+
+        if (!value.archive || !value.archive.restoreRequestedAt ||
+        !value.archive.restoreRequestedDays || isObjectAlreadyRestored) {
+            return;
+        }
+
+        // TODO: (To be double checked by the entry consumer)
```
Comment (nicolas2bert) [inline extensions/lifecycle/LifecycleQueuePopulator.js:150]: @RachedBenMustapha, could you please confirm?

File: extensions/lifecycle/tasks/LifecycleTask.js [hunk 2]
```diff
@@ -861,6 +861,44 @@ class LifecycleTask extends BackbeatTask {
      * @return {undefined}
      */
     _applyTransitionRule(params, log) {
+        // ZENKO-4173
+        // IF params.site is cold location:
+        // PUSH a Kafka entry to "cold-archive-req-<params.site>" topic
+        // Entry could look like:
+        // NOTE: the content of the entry should be review by the topic consumer's developer.
+        // [{
```
Comment (nicolas2bert) [inline extensions/lifecycle/tasks/LifecycleTask.js:869]: @RachedBenMustapha I came up with this format but we might already describe what `cold-archive-req-<params.site>` topic's consumer is expecting.

File: extensions/lifecycle/tasks/LifecycleTask.js [hunk 3]
```diff
@@ -861,6 +861,44 @@ class LifecycleTask extends BackbeatTask {
      * @return {undefined}
      */
     _applyTransitionRule(params, log) {
+        // ZENKO-4173
```
Comment (nicolas2bert) [inline extensions/lifecycle/tasks/LifecycleTask.js:864]: @naren-rajendran might be useful for your implementation.

File: extensions/lifecycle/LifecycleQueuePopulator.js [hunk 4]
```diff
@@ -117,6 +117,115 @@ class LifecycleQueuePopulator extends QueuePopulatorExtension {
...
+        if (operation !== 's3:ObjectRestore') {
+            return;
+        }
+
+        const locationName = value.dataStoreName;
+
+        // if object already restore nothing to do, S3 has already updated object MD
+        const isObjectAlreadyRestored = value.archive
+        && value.archive.restoreCompletedAt
+        && new Date(value.archive.restoreWillExpireAt) >= new Date(Date.now());
```
Comment (miniscruff) [inline extensions/lifecycle/LifecycleQueuePopulator.js:143]:
```suggestion
        && new Date(value.archive.restoreWillExpireAt) >= new Date();
```
new date returns the now date

File: extensions/lifecycle/LifecycleQueuePopulator.js [hunk 5]
```diff
@@ -117,6 +117,115 @@ class LifecycleQueuePopulator extends QueuePopulatorExtension {
...
+
+        if (!value.archive || !value.archive.restoreRequestedAt ||
+        !value.archive.restoreRequestedDays || isObjectAlreadyRestored) {
```
Comment (miniscruff) [inline extensions/lifecycle/LifecycleQueuePopulator.js:146]: Its not possible for restore requested days to be zero right? Just checking

File: extensions/lifecycle/LifecycleQueuePopulator.js (hunk 4 above)
Comment (nicolas2bert) [reply extensions/lifecycle/LifecycleQueuePopulator.js:143]: This PR is supposed to be for developers to know what is needed and where things are supposed to land.
I have no intention to merge it :)

File: extensions/lifecycle/LifecycleQueuePopulator.js (hunk 5 above)
Comment (nicolas2bert) [reply extensions/lifecycle/LifecycleQueuePopulator.js:146]: No, it should not be possible.

**Pull request closed without merging**